gr-blocks/python/blocks/qa_rotator_cc.py

```diff
@@ -0,0 +1,322 @@
+
+        # Post the phase increment control messages out of order
+        self._post_phase_inc_ctrl_msg(new_phase_incs[0], int(offsets[0]))
+        self._post_phase_inc_ctrl_msg(new_phase_incs[2], int(offsets[2]))
+        self._post_phase_inc_ctrl_msg(new_phase_incs[1], int(offsets[1]))
+
+        self.tb.run()
+
+        # Confirm they are received in order
+        self._assert_tags(new_phase_incs, offsets)
+
+    def test_duplicate_phase_inc_updates(self):
+        """Test multiple phase increment updates scheduled for the same sample
+
+        The rotator block applies all updates scheduled for the same sample
+        offset. Hence, only the last update shall take effect.
+
+        """
+        n_updates = 3
+
+        # Post the phase increment control messages
+        new_phase_incs = []
+        for i in range(n_updates):
+            new_phase_inc, \
+                offset, \
+                expected_samples = self._test_scheduled_phase_inc_update()
+            new_phase_incs.append(new_phase_inc)
+
+        self.tb.run()
+
+        # All "n_updates" tags are expected to be present
+        self._assert_tags(new_phase_incs, [offset] * n_updates)
+
+        # However, only the last update takes effect on the rotated samples
+        self.assertComplexTuplesAlmostEqual(self.sink.data(),
+                                            expected_samples,
+                                            places=4)
+
+    def test_phase_inc_update_out_of_range(self):
+        """Test phase increment update sent for an out-of-range offset"""
+        self._setUp(n_samples=2**16)
+        n_half_samples = int(self.n_samples / 2)
+
+        # Schedule a phase increment update whose offset is initially
+        # out-of-range due to being in the future
+        new_phase_inc = 2 * np.pi * 0.1
+        self._post_phase_inc_ctrl_msg(new_phase_inc, offset=n_half_samples)
+
+        # Run the flowgraph and wait until the rotator block does some work
+        self.tb.start()
+        while (self.rotator_cc.nitems_written(0) == 0):
+            pass
+
+        # The out-of-range update (scheduled for the future) should not have
+        # been applied yet, but it should not have been discarded either.
+        self._assert_tags([], [])
+
+        # Wait until at least the first half of samples have been processed
+        while (self.rotator_cc.nitems_written(0) < n_half_samples):
+            pass
+
+        # Now, schedule an update that is out-of-range due to being in the past
+        self._post_phase_inc_ctrl_msg(new_phase_inc, offset=0)
+
+        # And run the flowgraph to completion
+        self.tb.wait()
+
+        # The increment update initially scheduled for the future should have
+        # been applied when processing the second half of samples. In contrast,
+        # the update scheduled for the past should have been discarded.
+        self._assert_tags([new_phase_inc], [n_half_samples])
```

Comment:
I should note that this test may be slightly unreliable, given that it has a bit of racing. I couldn't think of a better way to implement it, though.

Essentially, I wanted to call the `_assert_tags` on L303 after the first `work` call on the rotator block. Also, I wanted to post a control message on L310 after processing half of the samples. So I added `while` loops, which are not perfect, but worked for me.

Let me know what you think.